### ISSUE TITLE
parallel.prange loops producing random results

### ISSUE THREAD

**Issue by ngoldbaum** [cython/cython#2316]
This was originally found by @AshKelly in the context of his GSOC project with yt.

Take a look at this repo: https://github.com/ngoldbaum/cython-issue

You can build the .pyx file with "pip install -e ."

If you're on a machine with OpenMP support (e.g. on Linux, not MacOS), you'll see that `fun_sum` returns random results while `fun_sum_scalar` returns consistent results. For reference, here's what `fun_sum` looks like:

```cython
cimport numpy as np
import numpy as np
cimport cython
from cython.parallel import prange

def fun_sum():

    cdef np.float64_t[:] result = np.zeros(1)
    cdef int j
    cdef np.float64_t a = 3.0

    with nogil:
        for j in prange(0, 10000):
            result[0] += a + j

    return np.array(result)
```

**Comment by ngoldbaum** [comment]
Yes, in the real example where this was originally found we're compositing data in parallel onto a 2D image array. We found we could trigger it with an array of shape `(1,)`, so we figured that was simplest for the bug report.

**Comment by ngoldbaum** [comment]
@scoder so @ashkelly has come up with a way for cython to generate code that avoids the race condition. It seems cython is incorrectly not marking the array as a parallel reduction target. See the description in https://github.com/ngoldbaum/cython-issue/pull/1 for details.

Unfortunately it looks like array reductions are only supported in relatively recent compilers (gcc 6 or newer, or any compiler that implements OpenMP 4.5 or newer). Out of curiosity what is the minimum supported OpenMP version for cython?

**Comment by ngoldbaum** [comment]
I guess this was already raised a year or two ago in #1504.

**Comment by scoder** [comment]
I'm not sure there is a minimum OpenMP version. There's at least support for OpenMP 3 and probably also 2. We could just generate a readable compile time `#error` when this feature is used with an older OMP version. Maybe also issue a Cython warning.

**Comment by scoder** [comment]
@bearnshaw provided a test case here: https://github.com/bearnshaw/mult-cython
I'll close #1504 to keep the discussion over here.

**Comment by scoder** [comment]
PR welcome, with an `#error` for older OMP versions as described above.

**Comment by lorentzenchr** [comment]
IIUC, such a PR would consist of 2 parts:
1. Check OpenMP version via `_OPENMP` macro variable (201511 corresponds to 4.5), see https://stackoverflow.com/questions/1304363/how-to-check-the-version-of-openmp-on-linux and https://www.openmp.org/specifications/.
2. Detect an addition assignment `+=` to a memoryview / array / pointer in a prange.

**Comment by da-woods** [comment]
Is there a cost to doing this? If someone did

```
for a in prange(arr.shape[0]):
    arr[a] += 5
```

(Or something similar) that'd probably get identified as a parallel reduction when it really isn't. I imagine that has a speed cost, and is probably a lot more common then the genuine reduction.

**Comment by scoder** [comment]
> ```
> for a in prange(arr.shape[0]):
>     arr[a] += 5
> ```
> that'd probably get identified as a parallel reduction when it really isn't

OTOH, it seems difficult to implement array reductions on user side *without* Cython language support. Assuming there is no easy work-around, I'd lean towards considering this a reduction, always. We document the in-place operators to mean a reduction inside of `prange` loops, so users should be aware of it and not be *too* surprised. I'd rather make it an error if an in-place operator usage *cannot* be implemented as a reduction by Cython, to make users aware of the difference and avoid unexpected behaviour.

**Comment by vovaf709** [comment]
So, are there any workarounds for array reductions in cython?
